### DISCUSSION TITLE
Help

### DISCUSSION THREAD

**Discussion by BENFREDJRima-34** [mesa/mesa#1539] [Q&A] [answered]
Can anyone help me? 
*I would like to get the agent neighbors from communication () method, but I'm getting below error 
AttributeError: type object 'Model' has no attribute  'grid'

`class MyAgent(Agent):
        `def __init__(self, unique_id,pos, model):
            super().__init__(unique_id, model)
           self.unique_id = unique_id
           self.pos = pos
          self.state = State.Not_adopter
          self.RA = random.random()
          self.Acc = random.random()
          self.Uncer = random.random()
          self.Infl = random.random()
          self.Compat= random.ranom()
       def communication(self):
             neighbors = self.model.grid.get_neighborhood(self.pos,include_center=False)
             adopter_neighbors = [
               agent
               for agent in self.model.grid.get_cell_list_contents(neighbors)
                if agent.state is State.Adopter
        ]`

   
`class MyModel(Model):
    """A model with some number of agents."""
    def __init__(self, N, width, height):
        self.num_agents = N
        self.grid = MultiGrid(width, height, True)
        self.schedule = RandomActivation(self)
        self.running = True 
        
        """Create agents"""
        for i in range(self.num_agents):
            a = MyAgent(i, self, Model)
            self.schedule.add(a)
            # Add the agent to a random grid cell
            x = self.random.randrange(self.grid.width)
            y = self.random.randrange(self.grid.height)
            self.grid.place_agent(a, (x, y))
    def step(self):
        self.schedule.step()`

`def agent_portrayal(agent):
    
    portrayal = {"Shape": "circle", "r": 0.5, "Filled": "true", "Layer": 0}

    if agent.state == State.Not_adopter:
        portrayal["Color"] = ["#FF0000", "#FF9999"]
        portrayal["stroke_color"] = "#00FF00"
    else:
        portrayal["Color"] = ["#0000FF", "#9999FF"]
        portrayal["stroke_color"] = "#000000"
    return portrayal

grid = CanvasGrid(agent_portrayal,10,10,500,500)
server= ModularServer(MyModel, [grid],
                       " Model", {"N":100,"width":10,"height":10}
                      )
server.port = 8521 # The default
server.launch()`

**Comment by rht** [comment, marked as answer]
```python
a = MyAgent(i, self, Model)
```

This should have been
```python
a = MyAgent(i, self)
```

You have to update the `__init__` definition of `MyAgent` to be consistent, by removing the `pos` from the argument. You don't have to assign `pos` to `a` during the initialization, since later `a.pos` will be assigned by `self.grid.place_agent(a, (x, y))`. And the model object to be passed is `self` instead of `Model`.

**Reply by BENFREDJRima-34** [reply]
OK, thank you @rht, it works now, but I have another error now in the same part 
`def communication(self):
neighbors = self.model.grid.get_neighborhood(self.pos,include_center=False)
adopter_neighbors = [
agent
for agent in self.model.grid.get_cell_list_contents(neighbors)
if agent.state is State.Adopter
]`
the eroor:
for agent in self.model.grid.get_cell_list_contents(neighbors)
  File "C:\ProgramData\Anaconda3\lib\site-packages\mesa\space.py", line 73, in wrapper
    return wrapped_function(*args)
  File "C:\ProgramData\Anaconda3\lib\site-packages\mesa\space.py", line 396, in get_cell_list_contents
    return list(self.iter_cell_list_contents(cell_list))
  File "C:\ProgramData\Anaconda3\lib\site-packages\mesa\space.py", line 585, in <genexpr>
    return itertools.chain.from_iterable(
TypeError: cannot unpack non-iterable MyAgent object

**Reply by rht** [reply]
I found an error in your code, which would have a different error message. In `self.model.grid.get_neighborhood(self.pos,include_center=False)`, you need to add a second positional argument (a boolean) after `self.pos`.

This version works on my machine:
```python
import mesa


class MyAgent(mesa.Agent):
    def __init__(self, unique_id, model):
        super().__init__(unique_id, model)
        self.unique_id = unique_id
        self.state = 1  # State.Not_adopter
        # self.RA = random.random()
        # self.Acc = random.random()
        # self.Uncer = random.random()
        # self.Infl = random.random()
        # self.Compat= random.ranom()

    def communication(self):
        neighbors = self.model.grid.get_neighborhood(
            self.pos, True, include_center=False
        )
        adopter_neighbors = [
            agent
            for agent in self.model.grid.get_cell_list_contents(neighbors)
            if agent.state == 0
        ]
        print(adopter_neighbors)

    def step(self):
        self.communication()


class MyModel(mesa.Model):
    """A model with some number of agents."""

    def __init__(self, N, width, height):
        self.num_agents = N
        self.grid = mesa.space.MultiGrid(width, height, True)
        self.schedule = mesa.time.RandomActivation(self)
        self.running = True

        """Create agents"""
        for i in range(self.num_agents):
            a = MyAgent(i, self)
            self.schedule.add(a)
            # Add the agent to a random grid cell
            x = self.random.randrange(self.grid.width)
            y = self.random.randrange(self.grid.height)
            self.grid.place_agent(a, (x, y))

    def step(self):
        self.schedule.step()


m = MyModel(20, 500, 500)
m.step()
```

**Reply by BENFREDJRima-34** [reply]
This works, Thank you
